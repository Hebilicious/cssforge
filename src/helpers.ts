/**
 * Validates a name to ensure it doesn't contain invalid characters.
 * Throws an error if the name is invalid.
 * @example
 * ```ts
 * validateName("valid-name"); // returns true
 * validateName("invalid.name"); // throws error
 * ```
 */
export function validateName(name: string): boolean {
  if (name.length === 0 || name.includes(".") || name === "value") {
    throw new Error(
      `Invalid name: ${name}. Names must be non-empty strings that do not contain periods.`,
    );
  }
  return true;
}

/**
 * Converts a pixel value to a rem value.
 * If the value is not in pixels, it returns the original value.
 * @example
 * ```ts
 * pxToRem({ value: "16px" }); // "1rem"
 * pxToRem({ value: "1rem" }); // "1rem"
 * pxToRem({ value: "32px", rem: 16 }); // "2rem"
 * ```
 */
export function pxToRem({ value, rem = 16 }: { value: string; rem?: number }): string {
  const pxMatch = value.endsWith("px");
  if (pxMatch) {
    const pxValue = parseFloat(value.slice(0, -2));
    const remValue = pxValue / rem;
    return `${remValue}rem`;
  }
  return value;
}
