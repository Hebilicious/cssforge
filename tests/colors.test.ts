import { assertEquals } from "@std/assert";
import { defineConfig, processColors } from "../src/mod.ts";
import { getLines } from "./helpers.ts";

Deno.test("processColors - converts hex to oklch", () => {
  const config = defineConfig({
    colors: {
      palette: {
        value: {
          coral: {
            100: { hex: "#FF7F50" },
            200: { hex: "#FF6347" },
          },
        },
      },
    },
  });

  const { css, resolveMap } = processColors(config.colors);
  const lines = getLines(css);
  assertEquals(
    lines[1].trim(),
    "--palette-coral-100: oklch(73.511% 0.16799 40.24666);",
  );
  assertEquals(
    lines[2].trim(),
    "--palette-coral-200: oklch(69.622% 0.19552 32.32143);",
  );
  assertEquals(Array.from(resolveMap.keys()), [
    "palette.value.coral.100",
    "palette.value.coral.200",
  ]);
});

Deno.test("processColors - handles different color formats", () => {
  const config = defineConfig(
    {
      colors: {
        palette: {
          value: {
            brand: {
              default: { hex: "#FF0000" },
              200: { rgb: [0, 255, 0] },
              300: { hsl: [240, 100, 50] },
              400: { oklch: "oklch(0.7 0.2 270)" },
            },
          },
        },
      },
    } as const,
  );
  const { css, resolveMap } = processColors(config.colors);
  const lines = getLines(css);

  assertEquals(
    lines[1].trim(),
    "--palette-brand-200: oklch(86.644% 0.29483 142.49535);",
  );
  assertEquals(
    lines[2].trim(),
    "--palette-brand-300: oklch(45.201% 0.31321 264.05202);",
  );
  assertEquals(lines[3].trim(), "--palette-brand-400: oklch(70% 0.2 270);");
  assertEquals(
    lines[4].trim(),
    "--palette-brand-default: oklch(62.796% 0.25768 29.23388);",
  );

  assertEquals(Array.from(resolveMap.keys()), [
    "palette.value.brand.200",
    "palette.value.brand.300",
    "palette.value.brand.400",
    "palette.value.brand.default",
  ]);
});

Deno.test("processColors - handles string values", () => {
  const config = defineConfig({
    colors: {
      palette: {
        value: {
          simple: {
            white: "oklch(100% 0 0)",
            black: "#000",
          },
        },
      },
    },
  });
  const { css, resolveMap } = processColors(config.colors);
  const lines = getLines(css);
  assertEquals(
    lines[1].trim(),
    "--palette-simple-white: oklch(100% 0 0);",
  );
  assertEquals(
    lines[2].trim(),
    "--palette-simple-black: oklch(0% 0 0);",
  );
  assertEquals(Array.from(resolveMap.keys()), [
    "palette.value.simple.white",
    "palette.value.simple.black",
  ]);
});

Deno.test("processColors - handles themes", () => {
  const config = defineConfig({
    colors: {
      palette: {
        value: {
          simple: {
            white: "oklch(100% 0 0)",
            black: { hex: "#000" },
          },
        },
      },
      theme: {
        value: {
          light: {
            background: {
              value: {
                primary: "var(--1)",
                secondary: "var(--2)",
              },
              variables: {
                1: "palette.value.simple.white",
                2: "palette.value.simple.black",
              },
            },
          },
        },
      },
    },
  });
  const { css, resolveMap } = processColors(config.colors);
  const lines = getLines(css);
  assertEquals(
    lines[5].trim(),
    "--theme-light-background-primary: var(--palette-simple-white);",
  );
  assertEquals(
    lines[6].trim(),
    "--theme-light-background-secondary: var(--palette-simple-black);",
  );
  assertEquals(Array.from(resolveMap.keys()), [
    "palette.value.simple.white",
    "palette.value.simple.black",
    "theme.value.light.background.primary",
    "theme.value.light.background.secondary",
  ]);
});

Deno.test("processColors - handles transparency", () => {
  const config = defineConfig({
    colors: {
      palette: {
        value: {
          alpha: {
            softGray1: "oklch(14.48% 0 0 / 12%)",
            softGray2: "oklch(14.48% 0 0 / 24%)",
          },
        },
      },
    },
  });
  const { css, resolveMap } = processColors(config.colors);
  const lines = getLines(css);
  assertEquals(
    lines[1],
    "--palette-alpha-softGray1: oklch(14.48% 0 0 / 12%);",
  );
  assertEquals(
    lines[2],
    "--palette-alpha-softGray2: oklch(14.48% 0 0 / 24%);",
  );
  assertEquals(Array.from(resolveMap.keys()), [
    "palette.value.alpha.softGray1",
    "palette.value.alpha.softGray2",
  ]);
});

Deno.test("processColors - generates gradient with color variables", () => {
  const config = defineConfig({
    colors: {
      palette: {
        value: {
          coral: {
            "50": { hex: "#FF7E60" },
            "90": { hex: "#FF7F40" },
            "100": { hex: "#FF6347" },
          },
          indigo: {
            "100": { hex: "#4F46E5" },
          },
        },
      },
      gradients: {
        value: {
          orangeGradient: {
            value: {
              primary: {
                value:
                  "linear-gradient(261.78deg, var(--1) 33.1%, var(--2) 56.3%, var(--3) 65.78%, var(--4) 84.23%)",
                variables: {
                  "1": "palette.value.coral.50",
                  "2": "palette.value.coral.90",
                  "3": "palette.value.coral.100",
                  "4": "palette.value.indigo.100",
                },
              },
            },
          },
        },
      },
    },
  });

  const result = processColors(config.colors);
  const lines = getLines(result.css);
  const expected = [
    "--gradients-orangeGradient-primary: linear-gradient(261.78deg, var(--palette-coral-50) 33.1%, var(--palette-coral-90) 56.3%, var(--palette-coral-100) 65.78%, var(--palette-indigo-100) 84.23%);",
  ].join("\n");

  assertEquals(lines.at(-1), expected);
  assertEquals(Array.from(result.resolveMap.keys()), [
    "palette.value.coral.50",
    "palette.value.coral.90",
    "palette.value.coral.100",
    "palette.value.indigo.100",
    "gradients.value.orangeGradient.primary",
  ]);
});

Deno.test("processColors - handles themes referencing gradients", () => {
  const config = defineConfig({
    colors: {
      palette: {
        value: {
          coral: {
            "50": { hex: "#FF7E60" },
          },
        },
      },
      gradients: {
        value: {
          orangeGradient: {
            value: {
              primary: {
                value: "linear-gradient(to right, var(--c1), var(--c2))",
                variables: {
                  "c1": "palette.value.coral.50",
                  "c2": "palette.value.coral.50",
                },
              },
            },
          },
        },
      },
      theme: {
        value: {
          light: {
            background: {
              value: {
                primary: "var(--grad)",
              },
              variables: {
                "grad": "gradients.value.orangeGradient.primary",
              },
            },
          },
        },
      },
    },
  });

  const { css, resolveMap } = processColors(config.colors);
  const lines = getLines(css);
  const lastLine = lines.pop();
  assertEquals(
    lastLine,
    "--theme-light-background-primary: var(--gradients-orangeGradient-primary);",
  );
  assertEquals(Array.from(resolveMap.keys()), [
    "palette.value.coral.50",
    "gradients.value.orangeGradient.primary",
    "theme.value.light.background.primary",
  ]);
});
