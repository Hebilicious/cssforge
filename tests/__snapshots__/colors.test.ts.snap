export const snapshot = {};

snapshot[`processColors - converts hex to oklch 1`] = `
"/* Palette */
/* coral */
--palette-coral-100: oklch(73.511% 0.16799 40.24666);
--palette-coral-200: oklch(69.622% 0.19552 32.32143);"
`;

snapshot[`processColors - converts hex to oklch 2`] = `
[
  [
    "palette.coral.100",
    {
      key: "--palette-coral-100",
      value: "oklch(73.511% 0.16799 40.24666)",
      variable: "--palette-coral-100: oklch(73.511% 0.16799 40.24666);",
    },
  ],
  [
    "palette.coral.200",
    {
      key: "--palette-coral-200",
      value: "oklch(69.622% 0.19552 32.32143)",
      variable: "--palette-coral-200: oklch(69.622% 0.19552 32.32143);",
    },
  ],
]
`;

snapshot[`processColors - handles different color formats 1`] = `
"/* Palette */
/* brand */
--palette-brand-200: oklch(86.644% 0.29483 142.49535);
--palette-brand-300: oklch(45.201% 0.31321 264.05202);
--palette-brand-400: oklch(70% 0.2 270);
--palette-brand-default: oklch(62.796% 0.25768 29.23388);"
`;

snapshot[`processColors - handles different color formats 2`] = `
[
  [
    "palette.brand.200",
    {
      key: "--palette-brand-200",
      value: "oklch(86.644% 0.29483 142.49535)",
      variable: "--palette-brand-200: oklch(86.644% 0.29483 142.49535);",
    },
  ],
  [
    "palette.brand.300",
    {
      key: "--palette-brand-300",
      value: "oklch(45.201% 0.31321 264.05202)",
      variable: "--palette-brand-300: oklch(45.201% 0.31321 264.05202);",
    },
  ],
  [
    "palette.brand.400",
    {
      key: "--palette-brand-400",
      value: "oklch(70% 0.2 270)",
      variable: "--palette-brand-400: oklch(70% 0.2 270);",
    },
  ],
  [
    "palette.brand.default",
    {
      key: "--palette-brand-default",
      value: "oklch(62.796% 0.25768 29.23388)",
      variable: "--palette-brand-default: oklch(62.796% 0.25768 29.23388);",
    },
  ],
]
`;

snapshot[`processColors - handles string values 1`] = `
"/* Palette */
/* simple */
--palette-simple-white: oklch(100% 0 0);
--palette-simple-black: oklch(0% 0 0);"
`;

snapshot[`processColors - handles string values 2`] = `
[
  [
    "palette.simple.white",
    {
      key: "--palette-simple-white",
      value: "oklch(100% 0 0)",
      variable: "--palette-simple-white: oklch(100% 0 0);",
    },
  ],
  [
    "palette.simple.black",
    {
      key: "--palette-simple-black",
      value: "oklch(0% 0 0)",
      variable: "--palette-simple-black: oklch(0% 0 0);",
    },
  ],
]
`;

snapshot[`processColors - handles themes 1`] = `
"/* Palette */
/* simple */
--palette-simple-white: oklch(100% 0 0);
--palette-simple-black: oklch(0% 0 0);
/* Themes */
/* Theme: light */
/* background */
--theme-light-background-primary: var(--palette-simple-white);
--theme-light-background-secondary: var(--palette-simple-black);"
`;

snapshot[`processColors - handles themes 2`] = `
[
  [
    "palette.simple.white",
    {
      key: "--palette-simple-white",
      value: "oklch(100% 0 0)",
      variable: "--palette-simple-white: oklch(100% 0 0);",
    },
  ],
  [
    "palette.simple.black",
    {
      key: "--palette-simple-black",
      value: "oklch(0% 0 0)",
      variable: "--palette-simple-black: oklch(0% 0 0);",
    },
  ],
  [
    "theme.light.background.primary",
    {
      key: "--theme-light-background-primary",
      value: "var(--palette-simple-white)",
      variable: "--theme-light-background-primary: var(--palette-simple-white);",
    },
  ],
  [
    "theme.light.background.secondary",
    {
      key: "--theme-light-background-secondary",
      value: "var(--palette-simple-black)",
      variable: "--theme-light-background-secondary: var(--palette-simple-black);",
    },
  ],
]
`;

snapshot[`processColors - handles transparency 1`] = `
"/* Palette */
/* alpha */
--palette-alpha-softGray1: oklch(14.48% 0 0 / 12%);
--palette-alpha-softGray2: oklch(14.48% 0 0 / 24%);"
`;

snapshot[`processColors - handles transparency 2`] = `
[
  [
    "palette.alpha.softGray1",
    {
      key: "--palette-alpha-softGray1",
      value: "oklch(14.48% 0 0 / 12%)",
      variable: "--palette-alpha-softGray1: oklch(14.48% 0 0 / 12%);",
    },
  ],
  [
    "palette.alpha.softGray2",
    {
      key: "--palette-alpha-softGray2",
      value: "oklch(14.48% 0 0 / 24%)",
      variable: "--palette-alpha-softGray2: oklch(14.48% 0 0 / 24%);",
    },
  ],
]
`;

snapshot[`processColors - generates gradient with color variables 1`] = `
"/* Palette */
/* coral */
--palette-coral-50: oklch(73.58% 0.16378 34.33822);
--palette-coral-90: oklch(73.341% 0.17338 44.64289);
--palette-coral-100: oklch(69.622% 0.19552 32.32143);
/* indigo */
--palette-indigo-100: oklch(51.057% 0.23005 276.96564);
/* Gradients */
/* orangeGradient */
--gradients-orangeGradient-primary: linear-gradient(261.78deg, var(--palette-coral-50) 33.1%, var(--palette-coral-90) 56.3%, var(--palette-coral-100) 65.78%, var(--palette-indigo-100) 84.23%);"
`;

snapshot[`processColors - generates gradient with color variables 2`] = `
[
  [
    "palette.coral.50",
    {
      key: "--palette-coral-50",
      value: "oklch(73.58% 0.16378 34.33822)",
      variable: "--palette-coral-50: oklch(73.58% 0.16378 34.33822);",
    },
  ],
  [
    "palette.coral.90",
    {
      key: "--palette-coral-90",
      value: "oklch(73.341% 0.17338 44.64289)",
      variable: "--palette-coral-90: oklch(73.341% 0.17338 44.64289);",
    },
  ],
  [
    "palette.coral.100",
    {
      key: "--palette-coral-100",
      value: "oklch(69.622% 0.19552 32.32143)",
      variable: "--palette-coral-100: oklch(69.622% 0.19552 32.32143);",
    },
  ],
  [
    "palette.indigo.100",
    {
      key: "--palette-indigo-100",
      value: "oklch(51.057% 0.23005 276.96564)",
      variable: "--palette-indigo-100: oklch(51.057% 0.23005 276.96564);",
    },
  ],
  [
    "gradients.orangeGradient.primary",
    {
      key: "--gradients-orangeGradient-primary",
      value: "linear-gradient(261.78deg, var(--palette-coral-50) 33.1%, var(--palette-coral-90) 56.3%, var(--palette-coral-100) 65.78%, var(--palette-indigo-100) 84.23%)",
      variable: "--gradients-orangeGradient-primary: linear-gradient(261.78deg, var(--palette-coral-50) 33.1%, var(--palette-coral-90) 56.3%, var(--palette-coral-100) 65.78%, var(--palette-indigo-100) 84.23%);",
    },
  ],
]
`;

snapshot[`processColors - handles themes referencing gradients 1`] = `
"/* Palette */
/* coral */
--palette-coral-50: oklch(73.58% 0.16378 34.33822);
/* Gradients */
/* orangeGradient */
--gradients-orangeGradient-primary: linear-gradient(to right, var(--palette-coral-50), var(--palette-coral-50));
/* Themes */
/* Theme: light */
/* background */
--theme-light-background-primary: var(--gradients-orangeGradient-primary);"
`;

snapshot[`processColors - handles themes referencing gradients 2`] = `
[
  [
    "palette.coral.50",
    {
      key: "--palette-coral-50",
      value: "oklch(73.58% 0.16378 34.33822)",
      variable: "--palette-coral-50: oklch(73.58% 0.16378 34.33822);",
    },
  ],
  [
    "gradients.orangeGradient.primary",
    {
      key: "--gradients-orangeGradient-primary",
      value: "linear-gradient(to right, var(--palette-coral-50), var(--palette-coral-50))",
      variable: "--gradients-orangeGradient-primary: linear-gradient(to right, var(--palette-coral-50), var(--palette-coral-50));",
    },
  ],
  [
    "theme.light.background.primary",
    {
      key: "--theme-light-background-primary",
      value: "var(--gradients-orangeGradient-primary)",
      variable: "--theme-light-background-primary: var(--gradients-orangeGradient-primary);",
    },
  ],
]
`;

snapshot[`processColors - handles gradients with conditions 1`] = `
"/* Palette */
/* coral */
--palette-coral-50: oklch(73.58% 0.16378 34.33822);
/* Gradients */
/* orangeGradient */
@media (prefers-color-scheme: dark) {
  --gradients-orangeGradient-primary: linear-gradient(to right, var(--palette-coral-50), var(--palette-coral-50));
}"
`;

snapshot[`processColors - handles gradients with conditions 2`] = `
[
  [
    "palette.coral.50",
    {
      key: "--palette-coral-50",
      value: "oklch(73.58% 0.16378 34.33822)",
      variable: "--palette-coral-50: oklch(73.58% 0.16378 34.33822);",
    },
  ],
  [
    "gradients.orangeGradient.primary",
    {
      key: "--gradients-orangeGradient-primary",
      value: "linear-gradient(to right, var(--palette-coral-50), var(--palette-coral-50))",
      variable: "--gradients-orangeGradient-primary: linear-gradient(to right, var(--palette-coral-50), var(--palette-coral-50));",
    },
  ],
]
`;

snapshot[`processColors - handles palette colors with conditions 1`] = `
"/* Palette */
/* background */
--palette-background-light: oklch(100% 0 0);
/* backgroundDark */
@media (prefers-color-scheme: dark) {
  --palette-backgroundDark-dark: oklch(0% 0 0);
}"
`;

snapshot[`processColors - handles palette colors with conditions 2`] = `
[
  [
    "palette.background.light",
    {
      key: "--palette-background-light",
      value: "oklch(100% 0 0)",
      variable: "--palette-background-light: oklch(100% 0 0);",
    },
  ],
  [
    "palette.backgroundDark.dark",
    {
      key: "--palette-backgroundDark-dark",
      value: "oklch(0% 0 0)",
      variable: "--palette-backgroundDark-dark: oklch(0% 0 0);",
    },
  ],
]
`;

snapshot[`processColors - handles themes with class condition 1`] = `
"/* Palette */
/* base */
--palette-base-primary: oklch(73.511% 0.16799 40.24666);
/* Themes */
/* Theme: dark */
.dark-theme {
  /* background */
  --theme-dark-background-primary: var(--palette-base-primary);
}"
`;

snapshot[`processColors - handles themes with class condition 2`] = `
[
  [
    "palette.base.primary",
    {
      key: "--palette-base-primary",
      value: "oklch(73.511% 0.16799 40.24666)",
      variable: "--palette-base-primary: oklch(73.511% 0.16799 40.24666);",
    },
  ],
  [
    "theme.dark.background.primary",
    {
      key: "--theme-dark-background-primary",
      value: "var(--palette-base-primary)",
      variable: "--theme-dark-background-primary: var(--palette-base-primary);",
    },
  ],
]
`;

snapshot[`processColors - handles theme variantNameOnly 1`] = `
"/* Palette */
/* simple */
--palette-simple-white: oklch(100% 0 0);
--palette-simple-black: oklch(0% 0 0);
/* Themes */
/* Theme: light */
/* background */
--primary: var(--palette-simple-white);
--secondary: var(--palette-simple-black);"
`;

snapshot[`processColors - handles theme variantNameOnly 2`] = `
[
  [
    "palette.simple.white",
    {
      key: "--palette-simple-white",
      value: "oklch(100% 0 0)",
      variable: "--palette-simple-white: oklch(100% 0 0);",
    },
  ],
  [
    "palette.simple.black",
    {
      key: "--palette-simple-black",
      value: "oklch(0% 0 0)",
      variable: "--palette-simple-black: oklch(0% 0 0);",
    },
  ],
  [
    "theme.light.background.primary",
    {
      key: "--primary",
      value: "var(--palette-simple-white)",
      variable: "--primary: var(--palette-simple-white);",
    },
  ],
  [
    "theme.light.background.secondary",
    {
      key: "--secondary",
      value: "var(--palette-simple-black)",
      variable: "--secondary: var(--palette-simple-black);",
    },
  ],
]
`;
